networking_aci/plugins/ml2/drivers/mech_aci/cobra_client.py

```diff
@@ -47,37 +49,50 @@ def wrapper(self, *args, **kwargs):
 
         try:
             # check if token is still valid
-            token_validity = self.mo_dir.session.refreshTime - time.time()
+            token_validity = (self.mo_dir.session.refreshTime or 0) - time.time()
             if token_validity < cfg.CONF.ml2_aci.reauth_threshold:
-                if token_validity > 1:
-                    LOG.debug("Session only valid for %ss, refreshing auth", token_validity)
-                    self.mo_dir.reauth()
-                else:
-                    LOG.info("Session timed out (%ss), triggering relogin", token_validity)
-                    self.login()
+                LOG.debug("Acquiring session refresh lock")
+                with _TOKEN_REFRESH_LOCK:
+                    if token_validity >= cfg.CONF.ml2_aci.reauth_threshold:
+                        LOG.debug("Session was already refreshed by another thread, continuing")
+                    elif token_validity > 1:
+                        LOG.debug("Session only valid for %ss, refreshing auth", token_validity)
+                        self.mo_dir.reauth()
+                    else:
+                        LOG.info("Session timed out (%ss), triggering relogin", token_validity)
+                        self.login()
+                LOG.debug("Releasing session refresh lock")
 
             return func(self, *args, **kwargs)
         except RETRY_EXCEPTIONS as e:
             msg = ("Try {}/{}: Call to {}() failed due to {}: {}"
                    .format(retry, max_retries, func.__name__, e.__class__.__name__, e))
 
+            trigger_login = True
             if isinstance(e, (CommitError, QueryError)):
                 if isinstance(e, CommitError) and e.error == 102:
-                    LOG.info("%s - sleeping and retrying to avoid race condition".format(e.reason))
+                    LOG.warning("%s - sleeping and retrying to avoid race condition", e.reason)
+                    trigger_login = False
                     time.sleep(1)
                 elif e.error == 403:
                     # relogin on error 403
                     pass
                 else:
                     # reraise
+                    LOG.exception("Commit failed")
                     raise e
 
             if retry < max_retries:
-                LOG.info("%s - calling login()", msg)
-                self.login()
-                return func(self, *args, retry=retry + 1, max_retries=max_retries, **kwargs)
+                if trigger_login:
+                    LOG.info("%s - calling login()", msg)
+                    LOG.debug("Acquiring session refresh lock for login")
+                    with _TOKEN_REFRESH_LOCK:
+                        self.login()
```

Comment:
Is it safe to login again, if another greenthread already logged in? `login()` doesn't check anything. Just does an unconditional login. Does that invalidate the currently-running requests with the old token? I'd assume it doesn't.

In theory, anything could have happened between our checks for login and acquiring the lock. X other greenthreads could have acquired the log since then. So would it make sense to check for validity again or is that just an optimization and it's best to play safe and login unconditionally?

Reply:
Yes, anything could happen inbetween. The `login()` might happen after another login, but I was not really sure what would be better. We end up in this case when the original login has timed out (which should not happen often due to the frequently running syncloop) or when something goes wrong, e.g. an APIC fails and throws SSL Errors at us. Then we would end up in a situation where the token would still be valid, but the APIC would no longer talk to us. I have not looked into cobra (v3) enough to see if they reset the refreshTime on a connection error and just went with the assumption they don't and the safest way would be to call `login()` agan in all cases where we aren't sure.

I'm not entirely sure that this is the best way, but at the time it felt like the best in terms of risk/work/reward. I am open to suggestions. :)

Reply:
Especially the SSL-Error case or failed APIC case makes it a good idea to call `login()`. Thanks.